pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -414,12 +417,11 @@ func (c *ClusterMemberController) getInternalIPForNode(node *corev1.Node) (strin
 	return internalIP, nil
 }
 
-// isURLMappedToMember returns true if the given peerURL matches any of the members' peerURLs
-func isURLMappedToMember(peerURL string, members []*etcdserverpb.Member) bool {
-	for _, m := range members {
-		if peerURL == m.PeerURLs[0] {
-			return true
-		}
+func (c *ClusterMemberController) Enqueue() {
```

Comment:
I think this can create a race condition. Imagine this controller is running every 10s on a different goroutine than the member lister controller.

The updates done in this Enqueue method might not be fully visible during the controller sync. I like this idea with the set, but you maybe just want to implement it in the boring way by listing all members when the controller sync runs and just reuse that return throughout the whole lifecycle of the sync.

OR alternatively add some locking around it, not sure it's really worth it because the controller runs never the less. If you were to tie the controller sync period to only run when the member listing changes, that would be useful but likely not necessary.

Reply:
I am a bit confused now :D 

So `Enqueue` is invoked by the `etcdmemberlistercontroller` when the member list changes only. it updates it own cache and then sync Listener's caches by invoking `Enqueue()` .. Moreover, Enqueue calls 
`(c *EtcdMemberListerController) GetMembers()`  which uses the lock before reading the cache and returning a copy to the Listener 

Listener's Sync will use the local updated cache to reconcile 

Where is the race condition ?

Reply:
simply put, you're adjusting the internal structure of `clustermembercontroller` here that runs on a different controller goroutine.

Enqueue is called from the `etcdmemberlistercontroller` controller goroutine, so you need to lock here in this controller as well if you want to do this listener-style updates. 

I'm sure if you were to run both controllers in goroutines in a test with `-race` it would flag that.

Reply:
i see it now, but what is the right way then, without using extra locking ? .. i think we should reduce this, or ?

Reply:
in `clustermembercontroller.sync` I would just call `getMembers()` as it was with the client. This is properly locked and you always get consistent data back for the lifetime of the sync call. We can add the notification fanciness later to only run the controller when the member list changes.

Reply:
yes I agree, I am also using `[]Member` instead of `map[string]Member` since the controller need not lookup by memberIP .. 

I am adding also `GetHealthyMember` and `GetUnhealthyMembers` so that other controllers need not to use the etcdclient directly for any member listing 

WIP .. :D